src/libstd/sys/common/mod.rs

```diff
@@ -74,6 +75,15 @@ pub fn mkerr_libc<T: Int>(ret: T) -> IoResult<()> {
     }
 }
 
+// unix has nonzero values as errors
+pub fn mkerr_libc_new<T: Int>(ret: T) -> io::Result<()> {
+    if ret != Int::zero() {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(())
+    }
+}
```

Comment:
I [also added this](https://github.com/rust-lang/rust/pull/21936/files#diff-28f2fd684ad47d385427678d96d2dcd4R182) as part of `std::fs`, so we may just want to be sure to coordinate on it.